modules/misc/listenbrainz.c

```diff
@@ -0,0 +1,522 @@
+            .on_current_media_changed = OnCurrentMediaChanged ,
+        };
+    static struct vlc_player_timer_cbs const timer_cbs =
+        {
+            .on_update = OnTimerUpdate ,
+            .on_discontinuity = OnTimerStopped ,
+        };
+
+    vlc_playlist_t *playlist = vlc_intf_GetMainPlaylist (p_intf);
+    p_sys->player = vlc_playlist_GetPlayer (playlist);
+
+    vlc_player_Lock (p_sys->player);
+    p_sys->player_listener = vlc_player_AddListener (p_sys->player , &player_cbs , p_intf);
+    vlc_player_Unlock (p_sys->player);
+
+    if (!p_sys->player_listener)
+        goto error;
+
+    p_sys->timer_listener = vlc_player_AddTimer (p_sys->player , VLC_TICK_FROM_SEC (1) , &timer_cbs , p_intf);
+    if (!p_sys->timer_listener)
+        goto error;
+
+    vlc_mutex_init (&p_sys->lock);
+    vlc_cond_init (&p_sys->wait);
+
+    if (vlc_clone (&p_sys->thread , Run , p_intf , VLC_THREAD_PRIORITY_LOW))
+        goto error;
+
+    return VLC_SUCCESS;
+
+error:
+    vlc_player_Lock (p_sys->player);
+    if (p_sys->player_listener)
+        vlc_player_RemoveListener (p_sys->player , p_sys->player_listener);
+    if (p_sys->timer_listener)
+        vlc_player_RemoveTimer (p_sys->player , p_sys->timer_listener);
+    vlc_player_Unlock (p_sys->player);
+    free (p_sys);
+    return VLC_EGENERIC;
+}
+
+static void Close (vlc_object_t *p_this)
+{
+    intf_thread_t *p_intf = ( intf_thread_t * ) p_this;
+    intf_sys_t *p_sys = p_intf->p_sys;
+
+    vlc_cancel (p_sys->thread);
+    vlc_join (p_sys->thread , NULL);
+
+    vlc_vector_clear (&p_sys->queue);
+    vlc_UrlClean (&p_sys->p_submit_url);
+
+    vlc_player_Lock (p_sys->player);
+    vlc_player_RemoveListener (p_sys->player , p_sys->player_listener);
+    vlc_player_Unlock (p_sys->player);
+
+    vlc_player_RemoveTimer (p_sys->player , p_sys->timer_listener);
+
+    free (p_sys);
+}
+
+static void *Run (void *data)
+{
+    intf_thread_t *p_intf = data;
+    int canc = vlc_savecancel ();
+    bool b_wait = 0;
+    char *request , *payload;
+
+    intf_sys_t *p_sys = p_intf->p_sys;
+
+    while (1)
+    {
+        vlc_restorecancel (canc);
+
+        if (b_wait)
+            vlc_tick_wait (vlc_tick_now () + VLC_TICK_FROM_SEC (60)); // wait for 1 min
+
+        vlc_mutex_lock (&p_sys->lock);
+        mutex_cleanup_push (&p_sys->lock) ;
+        while (p_sys->queue.size == 0)
+            vlc_cond_wait (&p_sys->wait , &p_sys->lock);
+        vlc_cleanup_pop ();
+
+        payload = PreparePayload (p_intf);
+        vlc_mutex_unlock (&p_sys->lock);
+        canc = vlc_savecancel ();
+
+        if (!payload)
+        {
+            msg_Warn (p_intf , "Error: Unable to generate payload");
+            break;
+        }
+
+        request = PrepareRequest (p_intf , payload);
+        if (!request)
+        {
+            msg_Warn (p_intf , "Error: Unable to generate request body");
+            break;
+        }
+
+        int i_status = SendRequest (p_intf , request);
```

Comment:
SendRequest() might be blocking since it's doing I/O operations.
There is mechanism to interrupt all VLC I/O functions with the vlc_interrupt.h API but it is quite difficult to understand.
You should do like modules/stream_filter/prefetch.c.

Create an interrupt context before creating the Run thread: `sys->interrupt = vlc_interrupt_create();`
Setup the interrupt context from the Run thread: `vlc_interrupt_set(sys->interrupt);`
Interrupt the Run thread from the Close function: ` vlc_interrupt_kill(sys->interrupt);`
Clean the interrupt context after this call ^ : `vlc_interrupt_destroy(sys->interrupt);`

Reply:
I think I have fixed this correctly now.

Reply:
Indeed, that's correct !